File: VolRangeForecaster/session_stats.py
#!/usr/bin/env python3
"""
Session Consumption Statistics

Fetches 5 years of H1 bars from Oanda and computes what fraction of the daily
H-L range each session (Asia, London) historically consumes.

Output: VolRangeForecaster/data/session_stats.json

Usage:
    python3 VolRangeForecaster/session_stats.py
    python3 VolRangeForecaster/session_stats.py --years 3
    python3 VolRangeForecaster/session_stats.py GOLD EURUSD   # specific instruments only

Session definitions (Europe/London local time):
    Asia    : 00:00 – 06:00
    London  : 08:00 – 13:00
"""

import argparse
import json
import os
import sys
import time
from datetime import datetime, timezone, timedelta
from pathlib import Path
from zoneinfo import ZoneInfo

import numpy as np
import requests

# ── Oanda config (mirrors fetch_m1_oanda.py) ──────────────────────────────────
OANDA_ENV  = os.environ.get("OANDA_ENV", "practice")
OANDA_BASE = (
    "https://api-fxpractice.oanda.com" if OANDA_ENV == "practice"
    else "https://api-fxtrade.oanda.com"
)
OANDA_KEY = os.environ.get(
    "OANDA_KEY",
    "12ede1dbab4361aa039831ec942603d2-b1691eab23c45cd97d0c1e7e63e9d5cc",
)

BARS_PER_REQUEST = 5000
OUTFILE = Path(__file__).parent / "data" / "session_stats.json"
LONDON_TZ = ZoneInfo("Europe/London")

# ── Instruments (same as vol_range_forecast.py) ───────────────────────────────
INSTRUMENTS = {
    "GOLD":   "XAU_USD",
    "NQ":     "NAS100_USD",
    "EURUSD": "EUR_USD",
    "GBPUSD": "GBP_USD",
    "USDJPY": "USD_JPY",
    "AUDUSD": "AUD_USD",
    "NZDUSD": "NZD_USD",
    "USDCAD": "USD_CAD",
    "USDCHF": "USD_CHF",
    "GBPJPY": "GBP_JPY",
}

# ── Session windows (London local hour, inclusive start, exclusive end) ────────
SESSIONS = {
    "asia":   (0,  6),   # 00:00–06:00 London
    "london": (8,  13),  # 08:00–13:00 London
    "ny":     (13, 21),  # 13:00–21:00 London  (US open → NY close)
}

# Minimum H1 bars required for a session to be included in stats
MIN_SESSION_BARS = 3
# Minimum H1 bars for a full trading day (skip gaps / holidays)
MIN_DAY_BARS = 12


# ── Oanda H1 fetch ─────────────────────────────────────────────────────────────

def _headers():
    if not OANDA_KEY:
        raise RuntimeError("OANDA_KEY env var not set")
    return {"Authorization": f"Bearer {OANDA_KEY}"}


def _fetch_chunk(instrument: str, from_dt: datetime, count: int = BARS_PER_REQUEST) -> list:
    url    = f"{OANDA_BASE}/v3/instruments/{instrument}/candles"
    params = {
        "granularity": "H1",
        "count":       count,
        "price":       "M",
        "from":        from_dt.strftime("%Y-%m-%dT%H:%M:%S.000000000Z"),
    }
    for attempt in range(4):
        try:
            r = requests.get(url, headers=_headers(), params=params, timeout=30)
            if r.status_code in (404, 422):
                return []
            r.raise_for_status()
            return [
                {
                    "open":  float(c["mid"]["o"]),
                    "high":  float(c["mid"]["h"]),
                    "low":   float(c["mid"]["l"]),
                    "close": float(c["mid"]["c"]),
                    "time":  datetime.fromisoformat(
                        c["time"].replace("Z", "+00:00").replace(".000000000", "")
                    ),
                }
                for c in r.json().get("candles", [])
                if c.get("complete", True) and c.get("mid")
            ]
        except requests.RequestException as exc:
            wait = 2 ** attempt
            print(f"  attempt {attempt + 1} failed: {exc} — retrying in {wait}s")
            time.sleep(wait)
    return []


def _fetch_all_h1(instrument: str, years: int) -> list:
    all_bars = []
    cursor   = datetime.now(timezone.utc) - timedelta(days=365 * years)
    now      = datetime.now(timezone.utc)

    while cursor < now:
        chunk = _fetch_chunk(instrument, cursor)
        if not chunk:
            cursor += timedelta(days=7)  # skip gaps (holidays / weekends)
            continue
        all_bars.extend(chunk)
        cursor = chunk[-1]["time"] + timedelta(hours=1)
        print(f"  {chunk[-1]['time'].strftime('%Y-%m-%d')}  {len(all_bars):,} bars", end="\r")
        if len(chunk) < BARS_PER_REQUEST:
            break  # caught up to now
        time.sleep(0.05)

    print()
    return all_bars


# ── Session consumption calculation ───────────────────────────────────────────

def _compute_stats(bars: list) -> dict:
    """
    Group H1 bars by London date.  For each date compute daily H-L and per-session
    H-L and abs(O-C), then return percentile stats of each as % of daily H-L.
    """
    by_date: dict[str, dict] = {}
    for bar in bars:
        london_dt = bar["time"].astimezone(LONDON_TZ)
        dk        = london_dt.date().isoformat()
        h         = london_dt.hour
        if dk not in by_date:
            by_date[dk] = {"all": [], "asia": [], "london": [], "ny": []}
        by_date[dk]["all"].append(bar)
        for sess, (h0, h1) in SESSIONS.items():
            if h0 <= h < h1:
                by_date[dk][sess].append(bar)

    buckets: dict[str, list] = {
        "asia_hl": [], "asia_oc": [],
        "london_hl": [], "london_oc": [],
        "ny_hl": [], "ny_oc": [],
    }

    for dk, grp in by_date.items():
        all_b = grp["all"]
        if len(all_b) < MIN_DAY_BARS:
            continue
        daily_hl = max(b["high"] for b in all_b) - min(b["low"] for b in all_b)
        if daily_hl < 1e-9:
            continue

        for sess_name in ("asia", "london", "ny"):
            sess_b = sorted(grp[sess_name], key=lambda b: b["time"])
            if len(sess_b) < MIN_SESSION_BARS:
                continue
            sess_hl = max(b["high"] for b in sess_b) - min(b["low"] for b in sess_b)
            sess_oc = abs(sess_b[-1]["close"] - sess_b[0]["open"])
            buckets[f"{sess_name}_hl"].append(sess_hl / daily_hl * 100)
            buckets[f"{sess_name}_oc"].append(sess_oc / daily_hl * 100)

    def _pct_stats(arr: list) -> dict | None:
        if not arr:
            return None
        a = np.array(arr)
        return {
            "p50":  round(float(np.percentile(a, 50)), 1),
            "p75":  round(float(np.percentile(a, 75)), 1),
            "mean": round(float(np.mean(a)), 1),
            "n":    len(arr),
        }

    return {
        "asia":      _pct_stats(buckets["asia_hl"]),
        "asia_oc":   _pct_stats(buckets["asia_oc"]),
        "london":    _pct_stats(buckets["london_hl"]),
        "london_oc": _pct_stats(buckets["london_oc"]),
        "ny":        _pct_stats(buckets["ny_hl"]),
        "ny_oc":     _pct_stats(buckets["ny_oc"]),
    }


# ── Text-format export (mirrors vol_range_forecast.py format_report) ──────────

_LW = 34

def _divider(name: str) -> str:
    prefix = f"──── {name} "
    return prefix + "─" * max(0, _LW - len(prefix))


def format_session_stats_block(instruments_stats: dict) -> str:
    """Returns the SESSION STATS text block for pasting into the Pine Script."""
    lines = [_divider("SESSION STATS"), ""]
    for name, stats in instruments_stats.items():
        if not stats or not stats.get("asia") or not stats.get("london"):
            continue
        a    = stats["asia"]
        a_oc = stats.get("asia_oc")
        l    = stats["london"]
        l_oc = stats.get("london_oc")
        ny   = stats.get("ny")
        ny_oc = stats.get("ny_oc")
        lines.append(name)
        lines.append(f"Asia range      : {a['p50']}% median · {a['p75']}% 75th")
        if a_oc:
            lines.append(f"Asia O-C        : {a_oc['p50']}% median · {a_oc['p75']}% 75th")
        lines.append(f"London range    : {l['p50']}% median · {l['p75']}% 75th")
        if l_oc:
            lines.append(f"London O-C      : {l_oc['p50']}% median · {l_oc['p75']}% 75th")
        if ny:
            lines.append(f"NY range        : {ny['p50']}% median · {ny['p75']}% 75th")
        if ny_oc:
            lines.append(f"NY O-C          : {ny_oc['p50']}% median · {ny_oc['p75']}% 75th")
        lines.append("")
    return "\n".join(lines)


# ── Entry point ───────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(description="Compute session consumption stats from Oanda H1 data")
    parser.add_argument("pairs",   nargs="*", help="Instrument names (default: all)")
    parser.add_argument("--years", type=int, default=5, help="Years of H1 history to fetch (default 5)")
    args = parser.parse_args()

    selected = [p.upper() for p in args.pairs] if args.pairs else list(INSTRUMENTS.keys())
    unknown  = [p for p in selected if p not in INSTRUMENTS]
    if unknown:
        print(f"Unknown instruments: {unknown}  (available: {list(INSTRUMENTS.keys())})")
        sys.exit(1)

    if not OANDA_KEY:
        print("Error: OANDA_KEY env var not set")
        sys.exit(1)

    results = {}

    for name in selected:
        oanda_sym = INSTRUMENTS[name]
        print(f"\n{'=' * 55}")
        print(f"  {name} ({oanda_sym})  —  {args.years}yr H1")
        print(f"{'=' * 55}")
        bars = _fetch_all_h1(oanda_sym, args.years)
        if not bars:
            print("  SKIPPED — no bars returned")
            continue
        stats = _compute_stats(bars)
        results[name] = stats
        for label, key in [("Asia H-L",   "asia"),   ("Asia O-C",   "asia_oc"),
                            ("London H-L", "london"), ("London O-C", "london_oc"),
                            ("NY H-L",     "ny"),     ("NY O-C",     "ny_oc")]:
            s = stats.get(key)
            if s:
                print(f"  {label:<12} P50={s['p50']}%  P75={s['p75']}%  (n={s['n']})")

    if not results:
        print("\nNo data computed — exiting without writing output.")
        sys.exit(1)

    output = {
        "ok":          True,
        "computed_at": datetime.now(timezone.utc).isoformat(),
        "years":       args.years,
        "instruments": results,
        "export_block": format_session_stats_block(results),
    }

    OUTFILE.parent.mkdir(parents=True, exist_ok=True)
    OUTFILE.write_text(json.dumps(output, indent=2))
    print(f"\n{'=' * 55}")
    print(f"Written  →  {OUTFILE}")
    print()
    print(output["export_block"])


if __name__ == "__main__":
    main()

File: js/sessionStats.js
/**
 * Session Consumption Statistics
 *
 * Fetches H1 bars from Oanda and computes what fraction of the daily H-L range
 * each session (Asia 00:00–06:00 London, London 08:00–13:00 London) historically
 * consumes.  Results are written to VolRangeForecaster/data/session_stats.json
 * and served via /api/session-stats.
 *
 * No Python dependency — runs entirely in Node.js using the same Oanda key that
 * the vol-forecast scheduler already uses.
 */

import fs   from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname  = path.dirname(fileURLToPath(import.meta.url));
const OUTFILE    = path.join(__dirname, '..', 'VolRangeForecaster', 'data', 'session_stats.json');

const OANDA_ENV  = process.env.OANDA_ENV || 'practice';
const OANDA_BASE = OANDA_ENV === 'practice'
  ? 'https://api-fxpractice.oanda.com'
  : 'https://api-fxtrade.oanda.com';

const BARS_PER_REQUEST = 5000;
const DEFAULT_YEARS    = 5;

// ── Instruments to analyse ────────────────────────────────────────────────────
// Key = name used in the Pine Script paste;  value = Oanda instrument symbol
const INSTRUMENTS = {
  GOLD:   'XAU_USD',
  NQ:     'NAS100_USD',
  EURUSD: 'EUR_USD',
  GBPUSD: 'GBP_USD',
  USDJPY: 'USD_JPY',
  AUDUSD: 'AUD_USD',
  NZDUSD: 'NZD_USD',
  USDCAD: 'USD_CAD',
  USDCHF: 'USD_CHF',
  GBPJPY: 'GBP_JPY',
};

// Session windows — London local time [inclusive start, exclusive end)
const SESSIONS = {
  asia:   [0,  6],   // 00:00–06:00
  london: [8,  13],  // 08:00–13:00
  ny:     [13, 21],  // 13:00–21:00  (US open → NY close)
};

const MIN_DAY_BARS     = 12;  // H1 bars required for a valid full trading day
const MIN_SESSION_BARS =  3;  // H1 bars required for a valid session slice


// ── Oanda H1 fetch (chunked pagination) ──────────────────────────────────────

function _oandaKey() {
  const key = process.env.OANDA_KEY;
  if (!key) throw new Error('OANDA_KEY env var not set');
  return key;
}

async function _fetchChunk(instrument, fromDate) {
  const url = new URL(`${OANDA_BASE}/v3/instruments/${instrument}/candles`);
  url.searchParams.set('granularity', 'H1');
  url.searchParams.set('count',       String(BARS_PER_REQUEST));
  url.searchParams.set('price',       'M');
  url.searchParams.set('from',        fromDate.toISOString().replace(/\.\d+Z$/, '.000000000Z'));

  for (let attempt = 0; attempt < 4; attempt++) {
    try {
      const res = await fetch(url.toString(), {
        headers: { Authorization: `Bearer ${_oandaKey()}` },
        signal:  AbortSignal.timeout(30_000),
      });
      if (res.status === 404 || res.status === 422) return [];
      if (!res.ok) throw new Error(`HTTP ${res.status}: ${await res.text()}`);

      const data = await res.json();
      return (data.candles ?? [])
        .filter(c => c.complete !== false && c.mid)
        .map(c => ({
          open:  parseFloat(c.mid.o),
          high:  parseFloat(c.mid.h),
          low:   parseFloat(c.mid.l),
          close: parseFloat(c.mid.c),
          time:  new Date(c.time),
        }));
    } catch (err) {
      if (attempt === 3) throw err;
      const wait = 2 ** (attempt + 1) * 1000;
      console.warn(`[SESSION-STATS] ${instrument} chunk attempt ${attempt + 1} failed: ${err.message} — retrying in ${wait / 1000}s`);
      await new Promise(r => setTimeout(r, wait));
    }
  }
  return [];
}

async function _fetchAllH1(instrument, years) {
  const all    = [];
  let   cursor = new Date(Date.now() - years * 365.25 * 24 * 60 * 60 * 1000);
  const now    = new Date();

  while (cursor < now) {
    const chunk = await _fetchChunk(instrument, cursor);
    if (!chunk.length) {
      // Non-trading gap (weekend / holiday) — skip forward 1 week
      cursor = new Date(cursor.getTime() + 7 * 24 * 60 * 60 * 1000);
      continue;
    }
    all.push(...chunk);
    cursor = new Date(chunk.at(-1).time.getTime() + 60 * 60 * 1000);
    process.stdout.write(`  [${instrument}] ${chunk.at(-1).time.toISOString().slice(0, 10)}  ${all.length.toLocaleString()} bars\r`);
    if (chunk.length < BARS_PER_REQUEST) break; // caught up to now
    await new Promise(r => setTimeout(r, 60));  // Oanda rate-limit buffer
  }
  process.stdout.write('\n');
  return all;
}


// ── Session consumption calculation ──────────────────────────────────────────

// Reusable London-time formatter
const _londonFmt = new Intl.DateTimeFormat('en-GB', {
  timeZone: 'Europe/London',
  year:     'numeric', month:  '2-digit', day:  '2-digit',
  hour:     '2-digit', hour12: false,
});

function _londonParts(date) {
  const parts = _londonFmt.formatToParts(date);
  const get = t => parts.find(p => p.type === t).value;
  return {
    date: `${get('year')}-${get('month')}-${get('day')}`,
    hour: parseInt(get('hour'), 10),
  };
}

function _computeStats(bars) {
  const byDate = new Map();

  for (const bar of bars) {
    const { date, hour } = _londonParts(bar.time);
    if (!byDate.has(date)) byDate.set(date, { all: [], asia: [], london: [], ny: [] });
    const grp = byDate.get(date);
    grp.all.push(bar);
    for (const [sess, [h0, h1]] of Object.entries(SESSIONS)) {
      if (hour >= h0 && hour < h1) grp[sess].push(bar);
    }
  }

  const buckets = {
    asia_hl: [], asia_oc: [],
    london_hl: [], london_oc: [],
    ny_hl: [], ny_oc: [],
  };

  for (const grp of byDate.values()) {
    if (grp.all.length < MIN_DAY_BARS) continue;
    const dailyHL = Math.max(...grp.all.map(b => b.high)) - Math.min(...grp.all.map(b => b.low));
    if (dailyHL < 1e-9) continue;

    for (const sess of ['asia', 'london', 'ny']) {
      const sb = grp[sess].slice().sort((a, b) => a.time - b.time);
      if (sb.length < MIN_SESSION_BARS) continue;
      const sessHL = Math.max(...sb.map(b => b.high)) - Math.min(...sb.map(b => b.low));
      const sessOC = Math.abs(sb.at(-1).close - sb[0].open);
      buckets[`${sess}_hl`].push(sessHL / dailyHL * 100);
      buckets[`${sess}_oc`].push(sessOC / dailyHL * 100);
    }
  }

  const _pct = (arr, p) => {
    if (!arr.length) return null;
    const s = [...arr].sort((a, b) => a - b);
    const i = (p / 100) * (s.length - 1);
    const lo = Math.floor(i), hi = Math.ceil(i);
    return Math.round((s[lo] + (s[hi] - s[lo]) * (i - lo)) * 10) / 10;
  };

  const mkStats = arr => arr.length
    ? { p50: _pct(arr, 50), p75: _pct(arr, 75), mean: Math.round(arr.reduce((s, v) => s + v, 0) / arr.length * 10) / 10, n: arr.length }
    : null;

  return {
    asia:      mkStats(buckets.asia_hl),
    asia_oc:   mkStats(buckets.asia_oc),
    london:    mkStats(buckets.london_hl),
    london_oc: mkStats(buckets.london_oc),
    ny:        mkStats(buckets.ny_hl),
    ny_oc:     mkStats(buckets.ny_oc),
  };
}


// ── Pine Script export block builder ─────────────────────────────────────────

const LW = 34;
function _div(name) { const p = `──── ${name} `; return p + '─'.repeat(Math.max(0, LW - p.length)); }

export function buildSessionExportBlock(instruments) {
  const lines = [_div('SESSION STATS'), ''];
  for (const [name, stats] of Object.entries(instruments)) {
    if (!stats?.asia || !stats?.london) continue;
    lines.push(name);
    lines.push(`Asia range      : ${stats.asia.p50}% median · ${stats.asia.p75}% 75th`);
    if (stats.asia_oc)   lines.push(`Asia O-C        : ${stats.asia_oc.p50}% median · ${stats.asia_oc.p75}% 75th`);
    lines.push(`London range    : ${stats.london.p50}% median · ${stats.london.p75}% 75th`);
    if (stats.london_oc) lines.push(`London O-C      : ${stats.london_oc.p50}% median · ${stats.london_oc.p75}% 75th`);
    if (stats.ny)        lines.push(`NY range        : ${stats.ny.p50}% median · ${stats.ny.p75}% 75th`);
    if (stats.ny_oc)     lines.push(`NY O-C          : ${stats.ny_oc.p50}% median · ${stats.ny_oc.p75}% 75th`);
    lines.push('');
  }
  return lines.join('\n');
}


// ── Public API ────────────────────────────────────────────────────────────────

let _computing = false;

/** Returns the cached stats from disk, or null if not yet computed. */
export function getSessionStats() {
  if (!fs.existsSync(OUTFILE)) return null;
  try { return JSON.parse(fs.readFileSync(OUTFILE, 'utf8')); }
  catch { return null; }
}

/** True while computeSessionStats() is running. */
export function isSessionStatsComputing() { return _computing; }

/**
 * Fetch H1 data from Oanda and compute session consumption stats.
 * Writes results to OUTFILE and returns the output object.
 * Throws if already computing.
 */
export async function computeSessionStats(years = DEFAULT_YEARS) {
  if (_computing) throw new Error('Session stats computation already in progress');
  _computing = true;
  const results = {};

  try {
    for (const [name, oandaSym] of Object.entries(INSTRUMENTS)) {
      console.log(`[SESSION-STATS] ── ${name} (${oandaSym}) ──`);
      const bars = await _fetchAllH1(oandaSym, years);
      if (!bars.length) { console.log('  SKIPPED — no bars'); continue; }
      const stats = _computeStats(bars);
      results[name] = stats;
      for (const [lbl, key] of [['Asia H-L','asia'],['Asia O-C','asia_oc'],['London H-L','london'],['London O-C','london_oc'],['NY H-L','ny'],['NY O-C','ny_oc']]) {
        const s = stats[key]; if (s) console.log(`  ${lbl.padEnd(12)} P50=${s.p50}%  P75=${s.p75}%  (n=${s.n})`);
      }
    }

    const output = {
      ok:           true,
      computed_at:  new Date().toISOString(),
      years,
      instruments:  results,
      export_block: buildSessionExportBlock(results),
    };

    fs.mkdirSync(path.dirname(OUTFILE), { recursive: true });
    fs.writeFileSync(OUTFILE, JSON.stringify(output, null, 2));
    console.log(`[SESSION-STATS] Written → ${OUTFILE}`);
    return output;
  } finally {
    _computing = false;
  }
}
